experimental/dds/tree2/src/util/utils.ts

```diff
@@ -373,6 +373,33 @@ export function objectToMap<MapKey extends string | number | symbol, MapValue>(
 	return map;
 }
 
+/**
+ * Convert an object used as a map into a new object used like a map..
+ *
+ * This function must only be used with objects specifically intended to encode map like information.
+ * The only time such objects should be used is for encoding maps as object literals to allow for developer ergonomics or JSON compatibility.
+ * Even those two use-cases need to be carefully considered as using objects as maps can have a lot of issues
+ * (including but not limited to unintended access to __proto__ and other non-owned keys).
+ * {@link objectToMap} helps these few cases get into using an actual map in as safe of was as is practical.
+ */
+export function transformObjectMap<MapKey extends string | number | symbol, MapValue, NewMapValue>(
+	objectMap: Record<MapKey, MapValue>,
+	transformer: (value: MapValue, key: MapKey) => NewMapValue,
+): Record<MapKey, MapValue> {
+	const output: Record<MapKey, MapValue> = Object.create(null);
+	// This function must only be used with objects specifically intended to encode map like information.
+	for (const key of Object.keys(objectMap)) {
+		const element = objectMap[key as MapKey];
+		Object.defineProperty(output, key, {
+			enumerable: true,
```

Comment:
Perhaps it makes sense to forward the enumerable, configurable, and writable flags from the original property? For example, if the input map's entry is not writeable, then it makes sense that the associated output entry should not be either. I guess enumerable will always be true here since you find the keys by enumerating.

Reply:
Since this creates a new object which the caller will have the only reference to, it makes sense for it to be mutable. If mutability isn't desired the caller can simply use TypeScript to restrict that (the reverse isn't valid, so this way is more general). Note that the TypeScript type returned requires mutability, so this should match that.

Reply:
Good point, that makes sense.